### ISSUE TITLE
Support AttributesGroupName with PreservedOrder for XmlBuilder

### ISSUE THREAD

**Issue by diomedtmc** [NaturalIntelligence/fast-xml-parser#420]
* [x] Are you running the latest version?
* [x] Have you included sample input, output, error, and expected output?
* [x] Have you checked if you are using correct configuration?
* [x] Did you try online [tool](https://naturalintelligence.github.io/fast-xml-parser/)?

### Description

It seems crazy to me that we could not build XML with an attributes group name and preserve order if we can parse it.  The framework already requires a special tag to set this up, and it includes a special tag to mark the group.  Why can't we just flatten the tree as we build if it is configured?

### Input
1.  Create a simple XML string with an attribute.
2.  Use XmlParser with preserveOrder = true and attributeGroupName = 'attributes'
3.  Use an XmlBuilder with the same values from  step 2 to regenerate the XML

### Code
```js
const sample = '<article><div style="font-family: 'Sans Serif';"><div></div><br/>Line 1<div></div></div></article>';
const sharedFastXmlConfiguration = {
    ignoreAttributes: false,
    attributesGroupName: AttributesGroupName,
    attributeNamePrefix: '',
    preserveOrder: true,
};

const parser = new XmlParser(sharedFastXmlConfiguration);
const builder = new XmlBuilder(sharedFastXmlConfiguration);
const elements = parser.parse(sample);
const sampleAfterFastXmlParser = builder.build(elements);

expect(sampleAfterFastXmlParser).to.equal(sample);
```

### Output

The values are not equal.  sampleAfterFastXmlParser will look like this:

```js
const sample = '<article><div attributes="[object Object]"><div></div><br/>Line 1<div></div></div></article>';
```

That's a crazy output.  When building the XML, and you encounter the key tag (":@") for an attribute group, pull out the nested attribute group name, and put all the attributes back into the element it belongs to.  Something like this:

```js
if (tag === ':@' && attributeGroupName) {
  for(const attributeKey in tag[attributeGroupName]) {
       element[attribute.subString(attributeNamePrefix.length)] = tag[attributeGroupName][attributeKey]
  }
}
```

### expected data
```
const sampleAfterFastXmlParser = '<article><div style="font-family: 'Sans Serif';"><div></div><br/>Line 1<div></div></div></article>';
```

**Would you like to work on this issue?**

- [ ] Yes
- [x] No

[Bookmark](https://github.com/NaturalIntelligence/fast-xml-parser/stargazers) this repository for further updates.

**Comment by amitguptagwl** [comment]
What is the benefit of using `attributeGroupName` with `preserveOrder : true`?

**Comment by diomedtmc** [comment]
We need to be able to preserve the order of elements, as it creates the order of a visual rendering (ie - HTML), but the order of the attributes does not matter (ie - CSS).  When modify the tree internally we use fast-xml-parser for an intermediate conversion, but then we eventually want to get back to HTML + CSS.  IMHO, preserve order is probably only useful for elements, and never attributes.  But having attributes parsed and grouped into a single named tag, highly useful when processing a tree (rather than looping over all tags to figure out which are attributes, which is the element name. etc... - we can just say "give me the attributes and let me modify those").

**Comment by amitguptagwl** [comment]
What I meant to ask is with `preserveOrder : true`, attributes are already grouped under `:@`. So what is the benefit to group them again? Probably I can change the code to rename `:@` to the `attributeGroupName` property. But I still doesn't see any advantage of doing that.

**Comment by diomedtmc** [comment]
It is not about grouping them on parse, it is if I do use a group on parse, when I call build with the same settings, they are not mapped back into the element:

```
const sample = '<article><div attributes="[object Object]"><div></div><br/>Line 1<div></div></div></article>';
```

I agree that order preservation of the attributes themselves is meaningless, but if I am using symmetric build/parse options, they should at least output.  Try my configuration from the ticket, and note the output on build.  Instead of getting the attributes back into the element, we get  [<attributeGroupName>] = "[object Object]", which means the code has only removed one layer, then just called toString() on the attributes group.

**Comment by diomedtmc** [comment]
(Note this only applies for XmlBuilder when using XmlParser settings preserveOrder=true + attributeGroupName="something")

**Comment by amitguptagwl** [comment]
Exactly, this is what I'm saying. Try the parser and builder both with `preserveOrder: true` and `attributeGroupName: false`. Then let me know if you're getting the desire output.

So, for example, in your case

```js

  const sample = `<article><div style="font-family: 'Sans Serif';"><div></div><br/>Line 1<div></div></div></article>`;
  const sharedFastXmlConfiguration = {
      ignoreAttributes: false,
      // attributesGroupName: AttributesGroupName,
      attributeNamePrefix: '',
      preserveOrder: true,
      processEntities: false,
  };

  const parser = new XMLParser(sharedFastXmlConfiguration);
  const builder = new XMLBuilder(sharedFastXmlConfiguration);
  const elements = parser.parse(sample);
  const sampleAfterFastXmlParser = builder.build(elements);

  expect(sampleAfterFastXmlParser).toEqual(sample); 
```

It gives almost the same result ie `<article><div style="font-family: 'Sans Serif';"><div></div><br></br>Line 1<div></div></div></article>` other than suppressing unpaired tag `br`.

**Comment by diomedtmc** [comment]
Yeah, that builds correctly, but identifying what is an attribute versus a tag is very inefficient.  With the group, we can say “anything that is :@ is an attribute, anything else is the tag”, otherwise, we have to parse and do more.   The irony here is how in agreement we are about all things, other than the fact the builder can’t build with attributeGroupName and preserveOrder, which seems very possible other than for the sake of argument.

**Comment by diomedtmc** [comment]
Simply put - preserveOrder should have no bearing on the attribute configuration.  I see no real world application where it makes sense, but attribute grouping by itself is highly value as is order preservation.  They are two independent features.

**Comment by amitguptagwl** [comment]
When you set `preserveOrder: true`, attributes are always grouped separately then tag's data under `:@` property so they can be referenced by the builder and js object prettier. So it is quick.

Check [this](https://github.com/NaturalIntelligence/fast-xml-parser/blob/7ab589190ea0b22f08b4b8ff6e9bdd68c6efef2e/src/xmlparser/node2json.js#L40) for your reference.

**Comment by diomedtmc** [comment]
Ah, ok.  Let me try again.

**Comment by diomedtmc** [comment]
I do get it grouped the way I want with those settings.  I'd say that intersection is attributeGroupName and preserveOrder is confusing, and maybe that is the feedback with this ticket.  It seems that it could be handled a little differently (maybe the attributeGroupName replaces the ':@'?).  Feel free to close and/or consider it feedback for improvement.
